tests/test_discussions.py

```diff
@@ -0,0 +1,270 @@
+        self.project.reload()
+
+        assert_in(unreg, get_unsubscribes(self.project))
+
+        unreg.register(username='billy@gmail.com', password='password1')
+        assert(unreg.is_registered)
+
+        self.project.reload()
+        assert_not_in(unreg, get_unsubscribes(self.project))
+
+
+    def test_disable_and_enable_project_discussions(self):
+        url = api_url_for('enable_discussions', pid=self.project._id)
+        payload = {}
+
+        assert_true(self.project.mailing_enabled)
+
+        self.app.delete(url, payload, auth=self.user.auth)
+        self.project.reload()
+        assert_false(self.project.mailing_enabled)
+
+        self.app.post(url, payload, auth=self.user.auth)
+        self.project.reload()
+        assert_true(self.project.mailing_enabled)
+
+class TestEmailRejections(OsfTestCase):
+
+    def setUp(self):
+        super(TestEmailRejections, self).setUp()
+        self.user = AuthUserFactory()
+        self.user.reload()
+        self.project = ProjectFactory(creator=self.user, parent=None)
+        # TODO this email will need to be updated when we start logging, since an actual message,
+        # and potentially other dictionary items, will become necessary
+        self.message = {
+            'To': '{}@osf.io'.format(self.project._id),
+            'From': self.user.email,
+            'subject': 'Hi, Friend!',
+            'stripped-text': 'Are you really my friend?'
+        }
+        self.post_url = api_url_for('route_message')
+
+    @mock.patch('website.mails.send_mail')
+    @mock.patch('website.project.mailing_list.send_messages')
+    def test_working_email(self, mock_send_list, mock_send_mail):
+        self.app.post(self.post_url, self.message)
+
+        mock_send_mail.assert_not_called()
+        # Due to unicode/str non-equality in assert_called_with:
+        assert_equal(mock_send_list.call_count, 1)
+        assert_equal(mock_send_list.call_args[0][0]._id, self.project._id)
+        assert_equal(mock_send_list.call_args[0][1]._id, self.user._id)
+        [assert_equal(mock_send_list.call_args[0][2][key], self.message[key])
+            for key in self.message.keys()]
+
+    @mock.patch('website.mails.send_mail')
+    @mock.patch('website.project.mailing_list.send_messages')
+    def test_email_from_non_registered_user(self, mock_send_list, mock_send_mail):
+        self.message['From'] = 'non-email@osf.fake'
+
+        self.app.post(self.post_url, self.message)
+
+        mock_send_mail.assert_called_with(
+            reason='no_user',
+            to_addr='non-email@osf.fake',
+            mail=mails.DISCUSSIONS_EMAIL_REJECTED,
+            target_address='{}@osf.io'.format(self.project._id),
+            user=None,
+            node_type='project',
+            node_url=self.project.absolute_url,
+            is_admin=False
+        )
+        mock_send_list.assert_not_called()
```

Comment:
Reminds me of http://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html.

And, actually, I don't find `assert_not_called` in http://www.voidspace.org.uk/python/mock/mock.html. Does that exist?